window-copy.c

```diff
@@ -3097,26 +3118,40 @@ window_copy_search(struct window_mode_entry *wme, int direction, int regex,
 	wrapflag = options_get_number(wp->window->options, "wrap-search");
 	cis = window_copy_is_lowercase(str);
 
-	if (direction)
+	/*
+	 * Behave according to mode-keys. If it is emacs, search forward leaves the
+	 * cursor after the match. If it is vi, the cursor remains at the beginning of
+	 * the match, regardless of direction.
+	 */
+	keys = options_get_number(wp->window->options, "mode-keys");
+	if (direction) {
+		if (keys == MODEKEY_VI) {
+			if (data->foundlen <= 0)
+				window_copy_move_right(s, &fx, &fy, wrapflag);
+			else
+				for (i = 0; i < data->foundlen; i++)
+					window_copy_move_right(s, &fx, &fy, wrapflag);
+		}
 		endline = gd->hsize + gd->sy - 1;
+	}
 	else {
-		if (again)
+		if (keys == MODEKEY_VI || again)
 			window_copy_move_left(s, &fx, &fy, wrapflag);
 		endline = 0;
 	}
 
 	found = window_copy_search_jump(wme, gd, ss.grid, fx, fy, endline, cis,
-	    wrapflag, direction, regex, &foundlen);
+	    wrapflag, direction, regex);
 	if (found) {
 		window_copy_search_marks(wme, &ss, regex, visible_only);
-		if (foundlen != 0) {
+		if (keys == MODEKEY_EMACS && data->foundlen != 0) {
 			/* Adjust for wrapped lines eating one right. */
-			i = data->cx + foundlen;
+			i = data->cx + data->foundlen;
 			while (i > gd->sx - 1) {
```

Comment:
Also, @nicm can you explain what's going on here, please?

Reply:
I didn't write this bit, @topcat001 did, he can probably tell you.

Reply:
When a line is wrapped, and you do a cursor right, then the cursor will stop at one past the end before moving on to the next line (it displays a '$' in that case). That's what I mean by "eating" one right. This means that if the line wraps n times, you need to insert n additional rights in order for the cursor to be positioned just after the match in this case. Does that help?